toktok/designs/osmock.md

```diff
@@ -0,0 +1,156 @@
+simulations and repeat recorded execution traces in toxcore.
+
+# Objective
+
+This design proposes to create a layer below toxcore containing mock versions
+of all system calls used by toxcore. Doing this gives us two advantages:
+
+1.  The operating system specific code is factored out, and none of the
+    implementation files `#include`s system headers other than libc and
+    library dependencies (opus, vpx).
+2.  None of the toxcore code performs any system calls directly, without going
+    through this layer.
+
+The first advantage allows us to test the OS-specific code independently from
+the portable code. Once we validated that the I/O code does I/O the right way,
+we can assume correctness and no longer need to test it through toxcore as
+thoroughly as before. We'll still want integration tests that test toxcore's
+integration with this layer, but most functional tests can become pure non-I/O
+tests. This will give an immense speedup of tests, improving developer
+velocity.
+
+The second point allows us to record exactly what system calls toxcore makes,
+together with the system's responses, in an execution trace. We can then later
+replay this trace to reproduce bugs. We can also use stored execution traces
+to ensure that toxcore behaves the same for the same input.
+
+Not only can we reproduce the exact same code execution, but we can also
+modify it slightly. This allows us to simulate arbitrary error conditions.
+Perhaps most interestingly, since `malloc` is potentially a system call, we
+will include it in the mock layer, allowing us to repeatedly execute the same
+trace, each time with an increasing number of successful mallocs. This allows
+us to test a large number of possible allocation failure handling paths, much
+more precisely than we would by just setting process resource limits.
+
+# Requirements
+
+Requirements for the solution:
+
+*   It should include *all* potential system calls, including libsodium/nacl
+    random functions such as key generation.
+*   It should not add more than 50ns of latency to actual system calls when
+    not recording or replaying a trace. The `gettimeofday` system call on
+    Linux takes about 50ns, so we take that as a benchmark.
+*   Recording a trace should be sufficiently low overhead such that we can
+    record the trace of a multimedia (A/V) call without noticable delay.
+*   It should be able to record toxcore API calls, so that we can replay the
+    activity of any API user without having to manually reproduce the API
+    calls that led to the execution.
+*   The remainder of the code, i.e. everything except the OS layer, should no
+    longer contain any platform-specific code. This includes LAN discovery
+    code.
+
+Specifically the LAN discovery code contains larger platform-specific
+functions that will need to move into the platform module.
+
+This proposal makes toxcore itself deterministic. The crypto library we use is
```

Comment:
"This proposal makes toxcore itself deterministic in tests"

Reply:
Done.